Add example for `Xterm.js` connected to Bash

paco-sevilla opened the pull request:
The discussion about integrating [Xterm.js](https://github.com/xtermjs/xterm.js) into `nicegui` (https://github.com/zauberzeug/nicegui/discussions/1846) seems to be stale...
However, I'd find it great to have `Xterm.js` in `nicegui`!

This PR just adds an example, but **I would be more than happy to close this PR and work on a proper integration** (including a "bell" event, a resizing event, addons like [@xterm/addon-web-links](https://github.com/xtermjs/xterm.js/tree/master/addons/addon-web-links), etc.).

## Use-cases for Xterm in nicegui:
* Provide a Terminal & Shell for native (electron-like) applications
* Provide a access to the Shell of a server through a web application, e.g. for a robot (this is the example in this PR, which has obvious security risks!)
* Display the output of a `subprocess` containing ANSI escape codes (this is what I'm mostly interested in)

## Example in this PR
The example integrates `Xterm.js` as a new element called `Terminal`, which provides a simple interface for reading and writing `bytes` from/to it.
This new element is then used in `main.py`, connecting in to `/bin/bash` running in a `pty.fork()` process.

The page looks as follows:
![image](https://github.com/user-attachments/assets/48c6a16d-1deb-4abf-824d-d02d3ee11dbe)

evnchn commented:
Xterm.js is quite a well-used project at this point. 

I have seen it used in Proxmox, as well as various third-party router plugins which provide SSH. 

So, I can stand behind including it for NiceGUI. 

---

If I understand correctly, now we need to do `npm install` to grab xterm right? 

Check out npm.json and npm.py to see how other elements officially in NiceGUI currently fetches the dependencies from NPM (though that's up for change later, but should stick until rest of 2.X)

paco-sevilla commented:
@evnchn Thanks for your comment.
I'm aware of those files (`npm.py` and `npm.json`). As I wrote above and in https://github.com/zauberzeug/nicegui/discussions/1846, if the authors of NiceGUI think that adding `Xterm.js` is valuable, I can work on a proper integration: add the module to the npm files, move `terminal.py` to `nicegui/elements`, add tests, ...

This PR just adds an example, very similar to the one here: https://github.com/zauberzeug/nicegui/tree/main/examples/signature_pad

rodja commented:
I think xterm.js would be a wonderful (and very powerful) NiceGUI element. But let me discuss it with @falkoschindler first. Might be at the end of the week.

falkoschindler commented:
Thanks for the pull request, @paco-sevilla!

@rodja and I agreed that it would be great to integrate Xterm.js in NiceGUI. For such a useful UI element, it only adds ~300KB of minified JS and a few lines of CSS. (I haven't checked the addons though.)

Instead of `ui.terminal`, we would prefer the name `ui.xterm`. For non-Quasar elements we tend to use the original brand name to avoid confusion with similar elements - like `ui.aggrid` was originally called `ui.table`, which is now a Quasar table...

So we'd appreciate it very much if you could convert the example into an integrated UI element, either on this PR or a new one. As @evnchn [pointed out](https://github.com/zauberzeug/nicegui/pull/4520#issuecomment-2745971673), we currently manage our dependencies with npm.json and npm.py, which might be a bit tricky to use. Let us know if you need help with anything!

KiddoV commented:
+1

evnchn commented:
Add this to `npm.json`

```json
  "xterm": {
    "package": "@xterm/xterm",
    "destination": "nicegui/elements/lib",
    "keep": [
      "package/lib/xterm.js",
      "package/css/xterm.css"
    ],
    "rename": {
      "package/lib/": "",
      "package/css/": ""
    },
    "version": "5.5.0"
  }
```

Shift things into `nicegui/elements/`, move to using `self.add_resource` for the CSS, and it should be much more better integrated.

paco-sevilla commented:
Sorry that I've been offline for so long... I finally found some time to work on this.
I've added the new element as requested, including tests, and updated the example.

I'll add some documentation/demos to the website in the next few days.
But I'm happy to receive a first pass of comments 🙂

paco-sevilla commented on .devcontainer/Dockerfile:
```diff
@@ -9,7 +9,7 @@ ENV POETRY_VERSION=1.6.1 \
 # Install packages
 RUN apt-get update && apt-get install --no-install-recommends -y \
-    sudo git build-essential chromium chromium-driver \
+    sudo git build-essential chromium chromium-driver chromium-sandbox \
```
This is now required to run tests in the devcontainer.

paco-sevilla commented on poetry.lock:
```diff
@@ -1,12 +1,11 @@
-# This file is automatically @generated by Poetry 2.1.2 and should not be changed by hand.
+# This file is automatically @generated by Poetry 1.6.1 and should not be changed by hand.
```
Poetry only supports Python 3.8 up to version `1.6.1`.
A lock file generated with a newer version produces warnings (e.g. in the devcontainer).

However, I'm happy to revert this change, if preferred.

paco-sevilla commented:
> I just had to remove the example audio, because the URL wasn't reachable anymore. We either need a different source or host the audio file ourselves.

I'm still able to download the audio (https://www.soundjay.com/buttons/beep-07a.mp3). It's just a short beeping sound (7kB).
If it's ok for you, we could include it (or something similar) in this repo and in NiceGUI. But I don't think the audio is that important.

> Now I'm pondering about the purpose of `ui.xterm`... The first demos are short and simple, but could seem like a poor copy of `ui.log` or `ui.codemirror`.

Yeah, I sort of see your point... What makes Xterm special is actually the parsing of ANSI escape codes (which is [not trivial](https://stackoverflow.com/a/59521115)).
So, I think most use cases of the Xterm element will involve a process/command running in the background (either a subprocess or a pty process). However, there are some caveats...

#### Caveats for subprocesses:
We could add a method like `run_subprocess` to the element, which takes the same args as `asyncio.create_subprocess_exec` but with `stdout` and `stderr` set to `subrpocess.PIPE` by default. The logic would be very similar to the last demo in this PR.

The problem is that most commands detect that `stdout`/`stderr` are not tty's and don't print outputs with ANSI codes, which makes having `Xterm` a bit useless (you might as well print the output to `ui.log`). This might surprise some users.
Many commands have a `--color` flag (like `ls --color=always`) to at least print the ANSI codes for text formatting, but they will still not print something like a progress bar.
 
Therefore, you actually want to use a PTY in most cases. But...

#### Caveats for pty:
The `pty` module in the standard library is _very_ basic (and has an awkward API). If you want to —for example— resize the terminal, you'll probably want to use a third-party package like [ptyprocess](https://github.com/pexpect/ptyprocess/) or [shellous](https://github.com/byllyfish/shellous).

Even worse, all of the above only work on Unix systems. If you're on Windows, you'll need to use a different package: [pywinpty](https://github.com/andfoy/pywinpty). To my knowledge, there is no cross-platform solution out there (see also https://github.com/python/cpython/issues/85829).

### Final thoughts
This comment is already quite long 🙈 My proposal is:

* **If it's acceptable to have different behavior on Windows vs. Unix:**
We can add a `run_subprocess` method that takes the same args as `asyncio.create_subprocess_exec` but by default sets `stdout` and `stderr` (and `stdin`?) to `subrpocess.PIPE` on Windows and to a PTY file descriptor that we get from [pty.openpty](https://docs.python.org/3/library/pty.html#pty.openpty) on Unix. Instead of a method, it could also be a separate `Process` class that `ui.xterm` can  `.bind()` to.

* **Otherwise**:
We keep the API minimalistic and let the user decide how to connect to the `Xterm` element.

In any case, I think the work should happen in a separate PR. The current API shouldn't be affected.

evnchn commented on nicegui/elements/xterm/xterm.py:
```diff
@@ -0,0 +1,118 @@
+        Note that you might also need to resize the backing pty to match the new size of the terminal.
+
+        :return: AwaitableResponse that, if awaited, returns ``None`` once the method call is finished.
+                 Otherwise, the method is executed without waiting for it to finish.
```
We do `:return: AwaitableResponse that can be awaited to get the result of the method call` for other elements, so I am thinking of `:return: AwaitableResponse that can be awaited to wait for the completion of the method call` to keep it one-line, concise, and consistent. It's a new example though.

falkoschindler replied:
I updated the four docstrings accordingly.

evnchn commented on nicegui/events.py:
```diff
@@ -396,6 +396,16 @@ class JsonEditorChangeEventArguments(UiEventArguments):
+
+@dataclass(**KWONLY_SLOTS)
+class XtermDataEventArguments(UiEventArguments):
+    data: str
```
Kinda asymmetric to not have the Bell when you have the Data. Maybe we can leave it.

falkoschindler reviewed (changes requested):
There are still some things to do:

- [x] I think the documentation should move to "Controls" (like `ui.codemirror`).
- [x] We need another bell sound because the current one takes seconds to load (at least here in Germany). Or maybe its easier to trigger some visual indicator like a notification.
- [x] The `print` call in one of the demo is invisible when run online. A notification or maybe a `ui.log` is better.
- [x] 3 out of 5 tests are red because they are missing a page function. In 3.0 we lost the ability to write tests without an explicit page.
- [x] `test_write` will probably be flaky because it relies on nicegui.io. We should use a local `@page` instead.
- [ ] The fit addon doesn't seem to work (test fails).

Therefore I'll move this PR to milestone 3.2.
